sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala

```diff
@@ -167,9 +160,20 @@ object GenerateUnsafeProjection extends CodeGenerator[Seq[Expression], UnsafePro
         }
     }
 
+    val writeFieldsCode = if (isTopLevel && row == null) {
```

Comment:
for top level struct, the input is `ctx.INPUT_ROW`, which may be null.

Reply:
BTW this is the only place the input can be null, otherwise it's always safe to replace the old `ctx.splitExpressions(row, ...)` with
```
ctx.splitExpressions(
       expressions = ...,
       funcName = ...,
       arguments = Seq("InternalRow" -> row))
```

Reply:
hmm, can the row be null here? I think `GenerateUnsafeProjection` is proposed to use to project against an input row.

Actually, I think it is confusing naming. `ctx.INPUT_ROW` is just used as the name of input row. It doesn't actually want to refer `ctx.INPUT_ROW` here.

```scala
public UnsafeRow apply(InternalRow ${ctx.INPUT_ROW}) {
  ${eval.code.trim}
  return ${eval.value};
}
```

Reply:
Also this codegen context in `GenerateUnsafeProjection` is newly created and used locally. I think the `ctx.INPUT_ROW` here can't be null.

```scala
private def create(
    expressions: Seq[Expression],
    subexpressionEliminationEnabled: Boolean): UnsafeProjection = {
  val ctx = newCodeGenContext()
  ...
```

Reply:
ideally the row can't be null, the top level is a special case that we may pass `ctx.INPUT_ROW` as input row.

Reply:
This codegen context isn't the same used in wholestage codegen.

Reply:
The `INPUT_ROW` of this local codegen context is not changed in `GenerateUnsafeProjection`. If it is `null`, the generated code will become:

```scala
class SpecificUnsafeProjection extends ${classOf[UnsafeProjection].getName} {
  ...
  public UnsafeRow apply(InternalRow null) {  // InternalRow ${ctx.INPUT_ROW} here.
    ${eval.code.trim}
    return ${eval.value};
  }
  ...
}
```

Reply:
I'm not sure I follow, if it's null, here I give up the splitting and just return `writeFields.mkString("\n")`. This is exactly what the previous code did

Reply:
maybe I should add one more check: `ctx.currentVars != null`

Reply:
`GenerateUnsafeProjection` is an compile unit by itself. The generated from it doesn't mix with wholestage codegen. It also has its codegen context. That is why I think this change doesn't make much sense. It also doesn't make sense to support wholestage codegen for it, IIUC.

Reply:
`GenerateUnsafeProjection.createCode` can be used in whole stage codegen

Reply:
Ah, I see. Then it makes sense.